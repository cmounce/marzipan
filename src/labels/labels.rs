use crate::{
    preprocess::peg::{Alt, And, Dot, EOF, Not, Opt, Parser, Rule, Star, Tag},
    star,
    world::Stat,
};

pub fn print_labels(b: &Stat) {
    let mut parser = Parser::new(&b.code);
    if !parse_program().parse(&mut parser) {
        eprintln!("Couldn't parse stat's code: {:?}", b.code);
        return;
    }

    for capture in parser.iter() {
        if capture.kind() == "label" {
            println!("- {}", capture.text());
        }
    }
}

fn parse_program() -> impl Rule {
    (Opt((parse_line, star!(("\n", parse_line)))), EOF)
}

fn parse_line() -> impl Rule {
    Alt((parse_label_line, parse_any_line))
}

fn parse_any_line() -> impl Rule {
    (star!(Not("\n"), Dot), eol)
}

fn parse_label_line() -> impl Rule {
    let allowed = Alt(('A'..='Z', 'a'..='z', '0'..='9', "_", "@", "~"));
    (":", Tag("label", star!(allowed)), eol)
}

fn eol() -> impl Rule {
    Alt((And("\n"), EOF))
}
