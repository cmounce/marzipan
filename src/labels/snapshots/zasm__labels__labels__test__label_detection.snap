---
source: src/labels/labels.rs
expression: result
---
'
'Find all the labels and their references
'
'Examples of labels
:(foo)
:(namespace~foo)
:(.local)
:(@)
'
'Test sends
#send [foo]
#send [self:foo]
#[self:foo]
#[@b]
#[.local]
#[foo.local]
'
'Test conditional sends
#if blocked n #send [foo]
#if blocked s #[foo]
#if blocked e send [foo]
#if blocked w [foo]
